File: frontend/src/utils/e2eCrypto.ts
/**
 * End-to-end encryption for relay WebSocket connections.
 *
 * Uses ECDH P-256 for key exchange and AES-GCM for authenticated encryption.
 * All operations use the Web Crypto API (hardware-accelerated on Apple devices).
 *
 * Protocol:
 *   1. Client sends CLIENT_HELLO (ECDH pubkey + supported ciphers)
 *   2. Server responds with SERVER_HELLO (ECDH pubkey + selected cipher)
 *   3. Both derive shared AES key via HKDF-SHA256
 *   4. All subsequent messages are AES-GCM encrypted
 *
 * Handshake message format (binary):
 *   [0x00 0xE2 0xEE 0x00] [version: 1] [type: 1] [payload...]
 *
 * Encrypted message format (binary, after handshake):
 *   [nonce: 12 bytes] [ciphertext + GCM tag: variable]
 */

// --- Constants ---

const E2E_MAGIC = new Uint8Array([0x00, 0xe2, 0xee, 0x00]);
const PROTOCOL_VERSION = 1;

const MSG_CLIENT_HELLO = 0x01;
// SERVER_HELLO type = 0x02 (checked via byte value in completeHandshake)

const CIPHER_AES_256_GCM = 0x01;
const CIPHER_AES_128_GCM = 0x02;

const SUPPORTED_CIPHERS = [CIPHER_AES_256_GCM, CIPHER_AES_128_GCM];

function cipherKeyBits(cipherId: number): number {
  return cipherId === CIPHER_AES_128_GCM ? 128 : 256;
}

// --- Binary helpers ---

function isHandshakeMessage(data: ArrayBuffer): boolean {
  if (data.byteLength < 6) return false;
  const v = new Uint8Array(data, 0, 4);
  return v[0] === 0x00 && v[1] === 0xe2 && v[2] === 0xee && v[3] === 0x00;
}

function buildHandshake(type: number, payload: Uint8Array): ArrayBuffer {
  const msg = new Uint8Array(6 + payload.length);
  msg.set(E2E_MAGIC, 0);
  msg[4] = PROTOCOL_VERSION;
  msg[5] = type;
  msg.set(payload, 6);
  return msg.buffer;
}

function encodeClientHello(pubKey: Uint8Array, clientRandom: Uint8Array): Uint8Array {
  const buf = new Uint8Array(1 + SUPPORTED_CIPHERS.length + 2 + pubKey.length + 32);
  let off = 0;
  buf[off++] = SUPPORTED_CIPHERS.length;
  for (const c of SUPPORTED_CIPHERS) buf[off++] = c;
  buf[off++] = (pubKey.length >> 8) & 0xff;
  buf[off++] = pubKey.length & 0xff;
  buf.set(pubKey, off);
  off += pubKey.length;
  buf.set(clientRandom, off);
  return buf;
}

function parseServerHello(payload: Uint8Array) {
  let off = 0;
  const selectedCipher = payload[off++];
  const pkLen = (payload[off] << 8) | payload[off + 1];
  off += 2;
  const serverPubKey = payload.slice(off, off + pkLen);
  off += pkLen;
  const serverRandom = payload.slice(off, off + 32);
  return { selectedCipher, serverPubKey, serverRandom };
}

// --- Nonce ---

function makeNonce(direction: number, counter: number): Uint8Array {
  const nonce = new Uint8Array(12);
  nonce[0] = direction;
  // 7-byte big-endian counter in bytes 1-7; bytes 8-11 stay zero
  for (let i = 6; i >= 0; i--) {
    nonce[7 - i] = (counter >> (i * 8)) & 0xff;
  }
  return nonce;
}

// --- E2EWebSocket wrapper ---

/**
 * Drop-in wrapper around a raw WebSocket that adds E2E encryption.
 *
 * After wrapping, all `send()` calls are encrypted and all incoming messages
 * are decrypted before being delivered to `onmessage`. During the handshake
 * phase, outgoing messages are buffered.
 *
 * Usage:
 *   const raw = new WebSocket(url);
 *   const ws = new E2EWebSocket(raw);
 *   // use ws.send(), ws.onmessage, ws.onopen, ws.onclose, ws.close() as normal
 */
export class E2EWebSocket {
  private ws: WebSocket;
  private aesKey: CryptoKey | null = null;
  private sendCounter = 0;
  private handshakeDone = false;
  private pendingSends: (string | ArrayBuffer | Uint8Array)[] = [];
  private clientKeyPair!: CryptoKeyPair;
  private clientRandom!: Uint8Array;

  // Public API matching WebSocket interface
  onopen: ((ev: Event) => void) | null = null;
  onmessage: ((ev: MessageEvent) => void) | null = null;
  onerror: ((ev: Event) => void) | null = null;
  onclose: ((ev: Event | CloseEvent) => void) | null = null;
  binaryType: BinaryType;

  get readyState(): number {
    return this.ws.readyState;
  }

  constructor(ws: WebSocket) {
    this.ws = ws;
    this.binaryType = ws.binaryType;

    ws.onopen = (ev) => {
      this.startHandshake();
      // Notify caller that WS is open (they can start sending — messages will be buffered)
      this.onopen?.(ev);
    };

    ws.onmessage = (ev) => {
      this.handleIncoming(ev);
    };

    ws.onerror = (ev) => {
      this.onerror?.(ev);
    };

    ws.onclose = (ev) => {
      this.onclose?.(ev);
    };
  }

  send(data: string | ArrayBuffer | ArrayBufferView): void {
    if (!this.handshakeDone) {
      // Buffer until handshake completes
      if (data instanceof ArrayBuffer) {
        this.pendingSends.push(data);
      } else if (ArrayBuffer.isView(data)) {
        this.pendingSends.push(
          (data.buffer as ArrayBuffer).slice(data.byteOffset, data.byteOffset + data.byteLength),
        );
      } else {
        this.pendingSends.push(data);
      }
      return;
    }
    this.encryptAndSend(data);
  }

  close(code?: number, reason?: string): void {
    this.ws.onclose = null;
    this.ws.close(code, reason);
  }

  // --- Internal ---

  private async startHandshake(): Promise<void> {
    try {
      this.clientKeyPair = await crypto.subtle.generateKey(
        { name: 'ECDH', namedCurve: 'P-256' },
        false,
        ['deriveKey', 'deriveBits'],
      );
      const pubRaw = await crypto.subtle.exportKey('raw', this.clientKeyPair.publicKey);
      this.clientRandom = crypto.getRandomValues(new Uint8Array(32));
      const hello = encodeClientHello(new Uint8Array(pubRaw), this.clientRandom);
      this.ws.send(buildHandshake(MSG_CLIENT_HELLO, hello));
    } catch (err) {
      console.error('[E2E] Handshake start failed:', err);
    }
  }

  private handleIncoming(ev: MessageEvent): void {
    const { data } = ev;

    // During handshake: intercept SERVER_HELLO
    if (!this.handshakeDone && data instanceof ArrayBuffer && isHandshakeMessage(data)) {
      this.completeHandshake(data);
      return;
    }

    // After handshake: all incoming data is encrypted binary
    if (this.handshakeDone && data instanceof ArrayBuffer) {
      this.decryptAndDeliver(data);
      return;
    }

    // Shouldn't happen after handshake, but pass through just in case
    this.onmessage?.(ev);
  }

  private async completeHandshake(data: ArrayBuffer): Promise<void> {
    try {
      const view = new Uint8Array(data);
      const payload = view.slice(6);
      const { selectedCipher, serverPubKey, serverRandom } = parseServerHello(payload);
      const keyBits = cipherKeyBits(selectedCipher);

      const serverKey = await crypto.subtle.importKey(
        'raw',
        (serverPubKey as Uint8Array<ArrayBuffer>).buffer,
        { name: 'ECDH', namedCurve: 'P-256' },
        false,
        [],
      );

      const sharedBits = await crypto.subtle.deriveBits(
        { name: 'ECDH', public: serverKey },
        this.clientKeyPair.privateKey,
        256,
      );

      const ikm = await crypto.subtle.importKey('raw', sharedBits, 'HKDF', false, ['deriveKey']);

      const transcript = new Uint8Array(64);
      transcript.set(this.clientRandom, 0);
      transcript.set(serverRandom, 32);
      const salt = new Uint8Array(await crypto.subtle.digest('SHA-256', transcript));

      this.aesKey = await crypto.subtle.deriveKey(
        { name: 'HKDF', hash: 'SHA-256', salt: salt as BufferSource, info: new TextEncoder().encode('tmuxdeck-e2e-v1').buffer as ArrayBuffer },
        ikm,
        { name: 'AES-GCM', length: keyBits },
        false,
        ['encrypt', 'decrypt'],
      );

      this.handshakeDone = true;

      // Flush buffered sends
      for (const msg of this.pendingSends) {
        this.encryptAndSend(msg);
      }
      this.pendingSends = [];
    } catch (err) {
      console.error('[E2E] Handshake completion failed:', err);
    }
  }

  private async encryptAndSend(data: string | ArrayBuffer | ArrayBufferView | Uint8Array): Promise<void> {
    if (!this.aesKey) return;
    try {
      let plainBuf: ArrayBuffer;
      if (typeof data === 'string') {
        plainBuf = new TextEncoder().encode(data).buffer as ArrayBuffer;
      } else if (data instanceof ArrayBuffer) {
        plainBuf = data;
      } else if (ArrayBuffer.isView(data)) {
        plainBuf = (data.buffer as ArrayBuffer).slice(data.byteOffset, data.byteOffset + data.byteLength);
      } else {
        plainBuf = data as ArrayBuffer;
      }

      const nonce = makeNonce(0x01, this.sendCounter++); // client direction
      const ct = await crypto.subtle.encrypt(
        { name: 'AES-GCM', iv: nonce as BufferSource },
        this.aesKey,
        plainBuf,
      );

      const frame = new Uint8Array(12 + ct.byteLength);
      frame.set(nonce, 0);
      frame.set(new Uint8Array(ct), 12);
      this.ws.send(frame.buffer as ArrayBuffer);
    } catch (err) {
      console.error('[E2E] Encrypt failed:', err);
    }
  }

  private async decryptAndDeliver(data: ArrayBuffer): Promise<void> {
    if (!this.aesKey) return;
    try {
      const view = new Uint8Array(data);
      const nonce = new Uint8Array(view.buffer, view.byteOffset, 12);
      const ct = new Uint8Array(view.buffer, view.byteOffset + 12);
      const plain = await crypto.subtle.decrypt(
        { name: 'AES-GCM', iv: nonce as BufferSource },
        this.aesKey,
        ct as BufferSource,
      );

      // Try decoding as text, fall back to binary
      let decoded: string | ArrayBuffer;
      try {
        decoded = new TextDecoder('utf-8', { fatal: true }).decode(plain);
      } catch {
        decoded = plain;
      }

      // Dispatch as a synthetic MessageEvent
      this.onmessage?.(new MessageEvent('message', { data: decoded }));
    } catch (err) {
      console.error('[E2E] Decrypt failed:', err);
    }
  }
}

/**
 * Detect if the current page is accessed through the relay (subdomain).
 * When accessed via relay, the hostname has the instance_id prefix, e.g.
 * "abc1234.relay.tmuxdeck.io" vs direct access at "localhost:3000".
 */
export function isRelayConnection(): boolean {
  const host = window.location.hostname;
  // Relay connections have at least 3 parts: instance_id.relay.domain.tld
  const parts = host.split('.');
  return parts.length >= 3 && !host.startsWith('www.');
}

File: backend/app/services/e2e_crypto.py
"""End-to-end encryption for relay WebSocket streams.

Uses ECDH P-256 for key exchange and AES-GCM for authenticated encryption.

Protocol:
  1. Client sends CLIENT_HELLO (ECDH pubkey + supported ciphers)
  2. Server responds with SERVER_HELLO (ECDH pubkey + selected cipher)
  3. Both derive shared AES key via HKDF-SHA256
  4. All subsequent messages are AES-GCM encrypted

Handshake message format (binary):
  [0x00 0xE2 0xEE 0x00] [version: 1] [type: 1] [payload...]

Encrypted message format (binary):
  [nonce: 12 bytes] [ciphertext + GCM tag: variable]
"""

from __future__ import annotations

import asyncio
import os
import struct
from hashlib import sha256

from cryptography.hazmat.primitives.asymmetric.ec import (
    ECDH,
    SECP256R1,
    EllipticCurvePublicKey,
    generate_private_key,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

# --- Constants ---

E2E_MAGIC = b"\x00\xe2\xee\x00"
PROTOCOL_VERSION = 1

MSG_CLIENT_HELLO = 0x01
MSG_SERVER_HELLO = 0x02

# Cipher IDs
CIPHER_AES_256_GCM = 0x01
CIPHER_AES_128_GCM = 0x02

# Server preference order (strongest first)
SERVER_CIPHER_PREFERENCE = [CIPHER_AES_256_GCM, CIPHER_AES_128_GCM]


def _cipher_key_bytes(cipher_id: int) -> int:
    return 16 if cipher_id == CIPHER_AES_128_GCM else 32


# --- Handshake parsing ---


def is_handshake_message(data: bytes) -> bool:
    return len(data) >= 6 and data[:4] == E2E_MAGIC


def _build_handshake(msg_type: int, payload: bytes) -> bytes:
    return E2E_MAGIC + bytes([PROTOCOL_VERSION, msg_type]) + payload


def _parse_client_hello(payload: bytes) -> tuple[list[int], bytes, bytes]:
    """Returns (supported_ciphers, client_pubkey, client_random)."""
    offset = 0
    cipher_count = payload[offset]
    offset += 1
    ciphers = list(payload[offset : offset + cipher_count])
    offset += cipher_count
    pubkey_len = struct.unpack(">H", payload[offset : offset + 2])[0]
    offset += 2
    pubkey = payload[offset : offset + pubkey_len]
    offset += pubkey_len
    client_random = payload[offset : offset + 32]
    return ciphers, pubkey, client_random


def _encode_server_hello(
    cipher_id: int, pubkey: bytes, server_random: bytes
) -> bytes:
    """Encode SERVER_HELLO payload."""
    return (
        bytes([cipher_id])
        + struct.pack(">H", len(pubkey))
        + pubkey
        + server_random
    )


# --- E2E Session ---


class E2ESession:
    """Encrypts/decrypts messages for a single WebSocket stream."""

    def __init__(self, aes_gcm: AESGCM):
        self._aes = aes_gcm
        self._send_counter = 0

    def _make_nonce(self, counter: int) -> bytes:
        """12-byte nonce: [direction:1][counter:7][zero:4].
        Server uses direction=0x02, client uses 0x01."""
        nonce = bytearray(12)
        nonce[0] = 0x02  # server direction
        counter_bytes = counter.to_bytes(7, "big")
        nonce[1:8] = counter_bytes
        return bytes(nonce)

    def encrypt(self, plaintext: bytes) -> bytes:
        """Encrypt plaintext, returns [nonce:12][ciphertext+tag]."""
        nonce = self._make_nonce(self._send_counter)
        self._send_counter += 1
        ct = self._aes.encrypt(nonce, plaintext, None)
        return nonce + ct

    def decrypt(self, data: bytes) -> bytes:
        """Decrypt [nonce:12][ciphertext+tag], returns plaintext."""
        nonce = data[:12]
        ciphertext = data[12:]
        return self._aes.decrypt(nonce, ciphertext, None)


def _handle_client_hello_sync(data: bytes) -> tuple[bytes, E2ESession]:
    """Synchronous implementation of the E2E handshake (CPU-intensive).

    Performs ECDH key generation, shared secret exchange, and HKDF derivation.
    Should be called via asyncio.to_thread() to avoid blocking the event loop.
    """
    payload = data[6:]  # skip magic + version + type
    client_ciphers, client_pubkey_raw, client_random = _parse_client_hello(
        payload
    )

    # Select strongest cipher both sides support
    selected = None
    for cipher in SERVER_CIPHER_PREFERENCE:
        if cipher in client_ciphers:
            selected = cipher
            break
    if selected is None:
        raise ValueError("No mutually supported cipher")

    # Generate ephemeral ECDH key pair
    private_key = generate_private_key(SECP256R1())
    public_key = private_key.public_key()

    # Export our public key in uncompressed point format (same as Web Crypto 'raw')
    server_pubkey_raw = public_key.public_bytes(
        Encoding.X962, PublicFormat.UncompressedPoint
    )

    # Import client's public key
    client_pubkey = EllipticCurvePublicKey.from_encoded_point(
        SECP256R1(), client_pubkey_raw
    )

    # ECDH shared secret
    shared_secret = private_key.exchange(ECDH(), client_pubkey)

    # Server random
    server_random = os.urandom(32)

    # Salt = SHA-256(client_random || server_random) for transcript binding
    transcript = client_random + server_random
    salt = sha256(transcript).digest()

    # Derive AES key via HKDF
    key_bytes = _cipher_key_bytes(selected)
    derived_key = HKDF(
        algorithm=SHA256(),
        length=key_bytes,
        salt=salt,
        info=b"tmuxdeck-e2e-v1",
    ).derive(shared_secret)

    aes = AESGCM(derived_key)
    session = E2ESession(aes)

    # Build SERVER_HELLO
    hello_payload = _encode_server_hello(
        selected, server_pubkey_raw, server_random
    )
    server_hello = _build_handshake(MSG_SERVER_HELLO, hello_payload)

    return server_hello, session


async def handle_client_hello(data: bytes) -> tuple[bytes, E2ESession]:
    """Process a CLIENT_HELLO message and return (SERVER_HELLO_bytes, session).

    Offloads CPU-intensive crypto operations to a thread to avoid blocking
    the event loop.
    """
    return await asyncio.to_thread(_handle_client_hello_sync, data)
